LibreNMS/Modules/Vlans.php

```diff
@@ -0,0 +1,264 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @link       http://librenms.org
+ *
+ * @copyright  2025 Peca Nesovanovic
+ * @author     Peca Nesovanovic <peca.nesovanovic@sattrakt.com>
+ */
+
+namespace LibreNMS\Modules;
+
+use App\Facades\PortCache;
+use App\Models\Device;
+use App\Models\PortVlan;
+use App\Models\Vlan;
+use App\Observers\ModuleModelObserver;
+use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\Log;
+use LibreNMS\DB\SyncsModels;
+use LibreNMS\Interfaces\Data\DataStorageInterface;
+use LibreNMS\Interfaces\Discovery\VlanDiscovery;
+use LibreNMS\Interfaces\Module;
+use LibreNMS\OS;
+use LibreNMS\Polling\ModuleStatus;
+use SnmpQuery;
+
+class Vlans implements Module
+{
+    use SyncsModels;
+
+    /**
+     * @inheritDoc
+     */
+    public function dependencies(): array
+    {
+        return ['ports'];
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function shouldDiscover(OS $os, ModuleStatus $status): bool
+    {
+        return $status->isEnabledAndDeviceUp($os->getDevice());
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function shouldPoll(OS $os, ModuleStatus $status): bool
+    {
+        return false;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function discover(OS $os): void
+    {
+        $vlans = new Collection;
+        $ports = new Collection;
+
+        $dot1dBasePortIfIndex = SnmpQuery::hideMib()->walk('BRIDGE-MIB::dot1dBasePortIfIndex')->table();
+        $dot1dBasePortIfIndex = $dot1dBasePortIfIndex['dot1dBasePortIfIndex'] ?? [];
+        $index2base = array_flip($dot1dBasePortIfIndex);
+
+        if ($os instanceof VlanDiscovery) {
+            $vlanData = $os->discoverVlans($dot1dBasePortIfIndex);
+        }
+
+        if (empty($vlanData)) {
+            $vlanData = $this->discoverVlans8021($os->getDevice(), $dot1dBasePortIfIndex);
+        }
+
+        if (empty($vlanData)) {
+            $vlanData = $this->discoverVlans($os->getDevice(), $dot1dBasePortIfIndex);
+        }
+
+        foreach ($vlanData['basic'] ?? [] as $key => $data) { // basic vlan info
+            //if vlan have no name, use generic 'VLAN x' where x is vlan_id
+            $data['vlan_name'] = (empty($data['vlan_name'])) ? 'VLAN ' . $data['vlan_vlan'] : $data['vlan_name'];
+            $vlans->push(new Vlan($data));
+        }
+
+        Log::info(PHP_EOL . 'Basic Vlan data:');
+        ModuleModelObserver::observe(\App\Models\Vlan::class);
+        $this->syncModels($os->getDevice(), 'vlans', $vlans);
+
+        foreach ($vlanData['ports'] ?? [] as $key => $data) {
+            $data['vlan'] = $data['vlan'] ?? 0;
```

Comment:
If these were already PortVlan object, we wouldn't need all these null checks here.